### PR TITLE
Update tensorify pass to specialize symfloats we didn't tensorify away

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138667

As discussed w/ @ezyang offline, one way to de-risk the `specialize_float=False` rollout is to specialize all backed symfloats that we fail to tensorify away. This diff does a few things:

1) It fixes a bug where item_memo gets dropped (due to incorrect epoch invalidation)
2) It updates the tensorify pass to do the backup specialization

This pass was originally part of the [PR](https://github.com/pytorch/pytorch/pull/137782) that flips `specialize_float=False` but we learned that the blast radius is simply too large. We've pivoted to a more milestone driven approach where we learn from the failures of the aforementioned PR and cherry pick fixes into main first. After this current PR lands our strategy is as follows:

1) Integrate turning off specialize float only in the automatic dynamic pass.
2) Put up a canary diff that only turns off specialize float in `backend=eager` mode to sniff out symfloat related bugs in dynamo due to code paths we previously never exercised.
3) Put up a canary diff that only turns off specialize float in `backend=aot_eager` mode to sniff out symfloat related bugs in aotautograd due to code paths we previously never exercised.

### CODE REVIEW

File: torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py [hunk 1]
```diff
@@ -130,9 +130,25 @@ def aot_dispatch_base_graph(
             mod_when_exporting_non_strict, assigned_buffers
         )
 
-    saved_updated_flat_args_subclasses_desugared = pytree.tree_map_only(
-        torch.Tensor, lambda t: t.detach(), updated_flat_args_subclasses_desugared
-    )
+    # TODO: Refactor the following code so detach() persists item_memo
```
Comment (bobrenjc93) [inline torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:133]: It's hard for me to estimate how deep or shallow this rabbit hole is. If reviewers have a more accurate estimate and can push me one way or another (include the refactor in this PR or not), do let me know!

Comment (ezyang) [reply torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:133]: You can do it on its own (and I'd recommend it), but basically if the detach implementation has some part that is in Python you're golden, and if it's entirely C++ give up and do something else lol
